### ISSUE TITLE
sidebar UI, wrong width calculation and a bit more

### ISSUE THREAD

**Issue by vetall812** [CleverRaven/Cataclysm-DDA#58991]
### Describe the bug

Use provided sidebar.

### Steps to reproduce

I've changed all "columns" to "minimum_columns" in my sidebar.

### Expected behavior

Should be columns with minimal width

### Versions and configuration

[crash.log](https://github.com/CleverRaven/Cataclysm-DDA/files/9047046/crash.log)
[!_sidebar.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/9047053/_sidebar.zip)

**Comment by Inglonias** [comment]
/confirmed

What's happening here is that in the attached sidebar, there is a layout called `a_safe_mode` with a width of 7. It has three widgets. The first widget has a width of 2. The second has a width of 7. Because of how minimum_columns works, the third widget in the list is given a width of -2, which makes the code very sad and causes the crash.

@bombasticSlacks

**Comment by bombasticSlacks** [comment]
sure I'll add a debug message at some point. You should just adjust the width though, I am just gonna make it throw a debug message and default to normal columns

**Comment by vetall812** [comment]
Okay. The best I can get is:
![image](https://user-images.githubusercontent.com/6676374/178497840-f69c0d7e-8d53-4cd4-91f1-08a76716f750.png)
the sidebar width is 65. 

Reducing width by one (64) gets following behavior:
![image](https://user-images.githubusercontent.com/6676374/178498201-107ea545-fb5f-4e70-8e2c-352e7f385260.png)
Notice "..." at the right border. Expected behavior is to reduce huge gaps between columns.

Continue reducing the width (63):
![image](https://user-images.githubusercontent.com/6676374/178498560-c1f319a4-7948-4932-9122-27d6d6503839.png)
This is an error about right arm widget.

Lets reduce width again (62):
![image](https://user-images.githubusercontent.com/6676374/178498802-19d871fc-e281-422e-88f9-05f760359d19.png)
Now error appears earlier - at the top part where are the main stats.

**Comment by Inglonias** [comment]
I'm not sure how many nested layouts you're using, but I suspect it's a lot, and I suspect that's not helping.

**Comment by bombasticSlacks** [comment]
I'm a bit confused as to what this sidebar should look like if designed correctly.

do you want there to not be the space between them (like I've drawn in red)
![image](https://user-images.githubusercontent.com/4514073/178501656-924c6036-2a03-462e-84b5-4594c6677803.png)

**Comment by vetall812** [comment]
yep, there should be no space here.

**Comment by vetall812** [comment]
> I'm a bit confused as to what this sidebar should look like if designed correctly.
> 
> do you want there to not be the space between them (like I've drawn in red) ![image](https://user-images.githubusercontent.com/4514073/178501656-924c6036-2a03-462e-84b5-4594c6677803.png)

![image](https://user-images.githubusercontent.com/6676374/178519012-638cf439-d763-4d7a-808a-26c63e8e52fb.png)
like this.

**Comment by vetall812** [comment]
> I'm not sure how many nested layouts you're using, but I suspect it's a lot, and I suspect that's not helping.

I've counted max depth is 5 (in first part of the sidebar)
So maximum depth is probably 7

![image](https://user-images.githubusercontent.com/6676374/178538974-da7a0498-537f-40b7-8a52-8749e4190283.png)

**Comment by bombasticSlacks** [comment]
can you post the JSON you are currently using and I'll see if I can play around with it a bit.

**Comment by vetall812** [comment]
> can you post the JSON you are currently using and I'll see if I can play around with it a bit.
[!_sidebar.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/9095525/_sidebar.zip)
here.

**Comment by vetall812** [comment]
> can you post the JSON you are currently using and I'll see if I can play around with it a bit.

Also, could you please also check - why is there an empty row after r_sidebar_row1, but no such empty row after r_sidebar_row2

**Comment by vetall812** [comment]
Oh, and I created alternate layout for bodyparts widgets. The one above is based on 4 rows, each one contains 3 meaningful columns.
Alt layout is based on 3 columns, each one contains 4 rows. Widgets are the same in each layout.
Alt layout failed to load. (
[!_sidebar_with_alt_layout.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/9095792/_sidebar_with_alt_layout.zip)

**Comment by vetall812** [comment]
After conversation with @bombasticSlacks and @dseguin the resulting picture is:
- There's a crash when using "minimum_columns"
- There's some padding issues when combining "padding" : 0 and the w_no_padding flag
- There's uncollapsible padding between column groups
- and all this somehow relates to width calculation and can be easily seen at widgets with 1 symbol width, such as "l_space" or "l_vc".

I've publish in this comment two JSONS:
1) all columns in sidebar are "arrange": "minimum_columns" and no "W_NO_PADDING" flag used at all.
[!_sidebar-minimum_columns.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/9123568/_sidebar-minimum_columns.zip)
the result of this JSON is:
![image](https://user-images.githubusercontent.com/6676374/179299362-280bf823-4cfd-44c2-a03d-4f7fbcfe3fbd.png)

2) all columns are simple "arrange": "columns", and there is the "W_NO_PADDING" flag set at the root of sidebar.
[!_sidebar-columns+no_padding.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/9123554/_sidebar-columns%2Bno_padding.zip)
this will lead to:
![image](https://user-images.githubusercontent.com/6676374/179299432-83a9b15c-20c9-4368-8fc0-02cba3797d28.png)

in both variants i've added alternate layout. The only difference is in order of nesting:
layout that active (r_sidebar_row2):
![image](https://user-images.githubusercontent.com/6676374/179299709-688aea91-0c94-450f-8576-fadad8285d55.png)

layout that exist in json but inactive (r_sidebar_row2_alt):
![image](https://user-images.githubusercontent.com/6676374/179299796-c6985dbc-9223-4e2e-9ace-3e2eeea0c81c.png)

this alt version calculated so wrong that i cannot run game with it. But i thought that they should be rendered the same result.

**Comment by vetall812** [comment]
It seems that the error comes from my templates.
I use the template "t_label" to define color for all labels on the sidebar.
All other labels are copy-from this template.
If I set width for any label equals to it actual width in symbols ("W:" = 2, "Light:" = 6) then something breaks and width became "undefined".
If I set width for the label as actual+1 - widget starts to use it, but design breaks.
@dseguin, may be it helps.

**Comment by vetall812** [comment]
I've tried to define all my widgets without templates - doesn't help.

Maybe the wrong calculation comes from label absent?

**Comment by dseguin** [comment]
So I gave this a try, and the crash when using `minimum_columns` can be fixed with:

```diff
diff --git a/src/widget.cpp b/src/widget.cpp
index 470a9ff8d3..ac92ff0d43 100644
--- a/src/widget.cpp
+++ b/src/widget.cpp
@@ -1576,7 +1576,7 @@ std::string widget::layout( const avatar &ava, unsigned int max_width, int label
                     }
                     // if last widget make it take the remaining space
                     if( wid == _widgets.back() ) {
-                        cur_width = avail_width - total_width;
+                        cur_width = std::max<int>( avail_width - total_width, 0 );
                     }
                 } else { //columns
                     if( cur_child._style == "layout" && cur_child._width > 1 ) {
```

Also, using the [minimum_columns sidebar](https://github.com/CleverRaven/Cataclysm-DDA/files/9123568/_sidebar-minimum_columns.zip), I was able to collapse the columns by just adding the `W_NO_PADDING` flag:

![col_fix](https://user-images.githubusercontent.com/12537966/179883427-67788b01-7411-4a0e-b63c-0475b184db37.png)

```diff
diff --git a/data/json/ui/_sidebar.json b/data/json/ui/_sidebar.json
index ba1ac510d0..4ddc15eec9 100644
--- a/data/json/ui/_sidebar.json
+++ b/data/json/ui/_sidebar.json
@@ -1663,6 +1663,7 @@
   { "id": "r_sidebar_row2",
     "style": "layout",  "arrange": "rows",
     "label": "! Player status",
+    "flags": [ "W_NO_PADDING" ],
     "widgets": [
       "a_wht_hd_pmf",
       "a_la_ct_ra",
```

**Comment by vetall812** [comment]
> ...
> Also, using the [minimum_columns sidebar](https://github.com/CleverRaven/Cataclysm-DDA/files/9123568/_sidebar-minimum_columns.zip), I was able to collapse the columns by just adding the `W_NO_PADDING` flag:
> 
> ![col_fix](https://user-images.githubusercontent.com/12537966/179883427-67788b01-7411-4a0e-b63c-0475b184db37.png)

As you can see - this flag brings another issue. It breaks right columns if left ones became thinner than expected. Take a look at the HD or RL widgets.

And this flag doesn't fix wrong width calculation. Examples:
- to part, right to the time is moon phase widget. It goes further than expected. 
- same moon phase widget added separately at the bottom part. There should be no space between left "bracket" and the moon. And there are others more subtle examples of course.

**Comment by mlange-42** [comment]
Maybe this is fixed by #59676?

**Comment by vetall812** [comment]
> Maybe this is fixed by #59676?

YES! thank you very much!
